packages/client/src/generation/utils.ts

```diff
@@ -344,3 +349,73 @@ export function unique<T>(arr: T[]): T[] {
 
   return result
 }
+
+/**
+ * Determines if arg types need generic <$PrismaModel> argument added.
+ * Essentially, performs breadth-first search for any fieldRefTypes that
+ * do not have corresponding `meta.source` defined.
+ *
+ * @param type
+ * @returns
+ */
+export function needsGenericModelArg(type: DMMF.SchemaArgInputType) {
+  const toVisit = [type]
+  const visited = new Set<string>()
+
+  let current: DMMF.SchemaArgInputType | undefined
+  while ((current = toVisit.shift())) {
+    const name = getTypeName(current)
+    if (visited.has(name)) {
+      // if the type is recursive, let other fields make a decision
+      continue
+    }
+    visited.add(name)
+
+    if (current.location === 'fieldRefTypes') {
+      return true
+    }
+
+    if (current.location === 'inputObjectTypes' && typeof current.type === 'object') {
```

Comment:
```suggestion
    if (current.location === 'inputObjectTypes') {
```

Would it always be an object?

Reply:
Thought so as well, but it some cases it is actually not: for example, [here](https://github.com/prisma/prisma/blob/2d36554ac7a537d082a405ac1d722d48efe86147/packages/client/src/generation/TSClient/Args.ts#L57) we create an input type manually and put a string there. There are couple of other similar places in generation code. Ideally, it would be nice to get rid of such cases and rely purely on `location` but I don't think it's feasible to do it in the scope of this task.